### ISSUE TITLE
Retroactively write a Materialized Views design document

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#16569]
For the benefit of new developers that want to enter the materialized view subsystem, we should have a **full**, **detailed** and **up-to-date** document which describes the design and implementation of this feature - with all its dozens of sub-features, cases, and so on.

We do already have a bunch of writeups, presentations, and so on, that several Scylla developers (Duarte, Sarna, I, etc.) wrote over the last 7 years, but none of it is actually full, detailed or up-to-date. 

The proposed design document should explain how the following things work, and how their code is structured. This is a long list, but probably still missing a lot of things:

* Basics of what MV does, how it's created, the wierd IS NOT NULL, etc.
* Outline of an update when views exist: row locking, read-before-write, etc.
* The hehavior difference between views that have a new key column and those they don't.
* Base-view pairing - why and how
* MV flow control - what appears to work well and what appears to be buggy.
* Filter on view's definition
* Secondary index
* Local Secondary index (a Scylla extension)
* Secondary index scanning
* Indexing collections
* Static columns in Index and MV
* Synchronous and asynchronous MV updates
* The behavior of empty keys
* Adding more than one new key column in a view (why it's not allowed and why it's still used in Alternator)
* MVs usage in Alternator
* Row liveness and the invisible "virtual columns"
* "shadowable row tombstone" - what it is and why we needed it
* The view building process (adding a view to an existing table) - how it distributes work, remembers progress, and completes.
* View-related metrics
* Whole-partition delete and its write amplification.
* View hints
* Views eventual consistency and repair (much of this is still open research, but there are some things we know)
* MV and SI tests - where are they
* MV-related system tables
* Calculated columns in MV, and how it's used (and why) for tokens in SI.
* CLUSTERING ORDER BY in materialized views
* view updates during range movements
* Backups and MV and SI
* Tablets and MV and SI

As you can see, this is a very long list of features, so even if we plan to redesign a few of these features in 2024, it is still likely we won't do a complete rewrite from scratch so many of these will remain relevant, so most of what we write in such a document would not go to waste even if pieces of MV are rewritten. 

It will be ideal to have such a document as an MD file in docs/dev, but it will be significantly easier to write it as a Google Docs document, so both should be considered.

CC @eliransin @kostja @mykaul @dorlaor @avikivity

**Comment by mykaul** [comment]
@kostja - can we start this effort, assuming we can incrementally complete it over time?

**Comment by kostja** [comment]
@mykaul I think the document written by Jan Ciolek is a good start https://docs.google.com/document/d/16yRwSQ2mNx9IuwLL4qMrws80xf08bx71LFvGlYHEOYg/edit#heading=h.qawplz88tny6

@piodul jfyi;

**Comment by kostja** [comment]
There are more documents available like https://docs.google.com/document/d/18c8JBS8-iY-X8s2EBm-uF8fXEabYHTSUAKVYpthhUgg/edit#heading=h.sctpfr1ygn03 and https://docs.google.com/document/d/1Kybb0PzQ9O7NSRS4_WcTdRyKXVBgJHRCGfsbtNCRFn8/edit#heading=h.7jls7hqk29bm but mostly they focus on issues, which we will address in one way or another.

**Comment by piodul** [comment]
Adding one more document to the pile: https://github.com/scylladb/scylladb/wiki/Materialized-Views . I encountered it last Friday. It looks like a general overview, but the last time it was updated was 5 years ago so it might be outdated in many aspects, and might not be as specific as we would want it to be (considering the huge list of items in this issue's description...).

**Comment by nyh** [comment]
Yes, as I already noted when I created the issue, we do have numerous documents and presentations on materialized views written over a period of over 7 years (much of it in the first year of development). At one point I made an effort to collected much of it, written by multiple developers, in the "Materialized Views and Secondary Index" of our shared "dev" Google drive (https://drive.google.com/drive/u/0/folders/1RaWjfjFjEpf765dLCwIoxEmfjDpoHghJ).

But writing the "design document" which I envisioned above will not be as simple as feeding all these documents into ChatGPT and asking it to summarize it into one document :-( These documents contain several good introductions to what Materialized Views is, but either contain very little concrete information on the dozens of separate subfeatures listed above (take a look at the list!) or contain outdated information or sometimes contradictory information.

**Comment by nyh** [comment]
> Adding one more document to the pile: https://github.com/scylladb/scylladb/wiki/Materialized-Views . I encountered it last Friday. It looks like a general overview, but the last time it was updated was 5 years ago so it might be outdated in many aspects, and might not be as specific as we would want it to be (considering the huge list of items in this issue's description...).

I wrote this document in 2016, which was 8 years ago, not 5 years ago. This was just before we implemented materialized views in Scylla, and described **Cassandra's** materialized views, not ours (take a look - it only mentions "Cassandra" by name, and refers to Cassandra source files!). Our implementation followed Cassandra's closely, but not everything is the same and almost everything in the long list of features I mentioned in the top of this issue isn't even mentioned. In the "write" section of that document, the details (order, batchlog, etc.) do **not** match what we have, and I'm not sure it's because it was always wrong, or maybe Cassandra changed since I wrote it, or maybe we changed.

**Comment by nyh** [comment]
I started working on a new document https://docs.google.com/document/d/1yk7MVkXM3LEDV_sb9FMHCQmMJYYwGN6fwsz3YzbOEc0/edit#heading=h.95xr14s9j0o9 - I created it as a Google Doc to make it easy and fast to edit, but because Scylla's Google Docs can no longer be visible outside the company, when the content will be complete it will have to be extracted into a file and put into our git repository. But until then, I plan to work on it in Google Docs.

This is just the beginning, and at the time of this writing the linked document has almost no content. It will probably take a long time to write the whole document, so please don't hold your breath :-)
